File: apps/backend/app/schemas/race.py
"""Response DTOs for `/races*` (docs/08_API_SPECIFICATION.md — "Races").

"Race" identifiers are `dim_session.session_id` UUIDs, scoped to
`session_type == "R"` rows — races aren't a separate Gold entity, they're
race sessions (see `app/models/gold/session.py`).

`RaceWeather` is a single per-session snapshot, not a time series: docs/08
calls this "Weather evolution", but `dim_weather` is deliberately one
aggregated row per session (see `app/models/gold/weather.py`) — there is no
FastF1-sample-level time series retained past the Gold transform.
`RaceStrategyEntry` is derived from tyre-compound *changes* across
`fct_laps` (not `fct_pitstop`), since Phase 4's pit-stop detection only
catches stops recorded on a single lap row (see `app/models/gold/pitstop.py`)
and would otherwise leave this endpoint mostly empty.
"""

from __future__ import annotations

import uuid

from app.schemas.base import CamelModel


class RaceListItem(CamelModel):
    id: uuid.UUID
    season: int
    round: int
    race_name: str | None
    circuit: str | None
    country: str | None
    date: str | None
    winner: str | None


class RaceSummary(CamelModel):
    id: uuid.UUID
    season: int
    round: int
    race_name: str | None
    circuit: str | None
    country: str | None
    date: str | None
    winner: str | None
    pole: str | None
    fastest_lap: str | None
    retirements: int | None
    weather: str | None
    average_pitstop: float | None


class PositionEntry(CamelModel):
    driver: str
    lap_number: float
    position: float | None


class PitstopEntry(CamelModel):
    driver: str
    lap: float
    pit_duration: float | None
    stop_number: int | None
    compound_before: str | None
    compound_after: str | None


class RaceWeather(CamelModel):
    air_temperature: float | None
    track_temperature: float | None
    humidity: float | None
    wind_speed: float | None
    wind_direction: float | None
    rainfall: bool | None
    pressure: float | None


class TyreStint(CamelModel):
    compound: str | None
    start_lap: float
    end_lap: float
    lap_count: int


class DriverStrategy(CamelModel):
    driver: str
    stints: list[TyreStint]

File: packages/shared-types/src/domain/race.ts
/**
 * `/races*` (docs/08_API_SPECIFICATION.md — "Races"). Mirrors
 * `apps/backend/app/schemas/race.py` exactly. A "race" is a
 * `dim_session` row scoped to `session_type == "R"` — not a separate
 * Gold entity — so `RaceListItem.id`/`RaceSummary.id` are the same
 * `session_id` UUIDs `/sessions/{id}` uses.
 *
 * `RaceWeather` is a single per-session snapshot, not a time series:
 * docs/08 calls this "weather evolution", but `dim_weather` is
 * deliberately one aggregated row per session — there is no
 * FastF1-sample-level time series retained past the Gold transform.
 */

export interface RaceListItem {
  id: string;
  season: number;
  round: number;
  raceName: string | null;
  circuit: string | null;
  country: string | null;
  date: string | null;
  winner: string | null;
}

export interface RaceSummary {
  id: string;
  season: number;
  round: number;
  raceName: string | null;
  circuit: string | null;
  country: string | null;
  date: string | null;
  winner: string | null;
  pole: string | null;
  fastestLap: string | null;
  retirements: number | null;
  weather: string | null;
  averagePitstop: number | null;
}

export interface PositionEntry {
  driver: string;
  lapNumber: number;
  position: number | null;
}

export interface PitstopEntry {
  driver: string;
  lap: number;
  pitDuration: number | null;
  stopNumber: number | null;
  compoundBefore: string | null;
  compoundAfter: string | null;
}

export interface RaceWeather {
  airTemperature: number | null;
  trackTemperature: number | null;
  humidity: number | null;
  windSpeed: number | null;
  windDirection: number | null;
  rainfall: boolean | null;
  pressure: number | null;
}

export interface TyreStint {
  compound: string | null;
  startLap: number;
  endLap: number;
  lapCount: number;
}

export interface DriverStrategy {
  driver: string;
  stints: TyreStint[];
}
